querying a range that includes the past and the future causes a panic

the following script will cause a panic.  the script is similar to the one added in #2159 with an extra `time > now() - 2d` in the where clause of the last query.

```sh
now=$(date '+%FT%T.000Z')
tomorrow=$(date -v +1d '+%FT%T.000Z')

echo "creating database"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "q=CREATE DATABASE foo"

echo "creating retention policy"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "q=CREATE RETENTION POLICY bar ON foo DURATION INF REPLICATION 1 DEFAULT"

echo "inserting data"
curl -d "{\"database\" : \"foo\", \"retentionPolicy\" : \"bar\", \"points\": [{\"name\": \"cpu\", \"tags\": {\"region\":\"uswest\",\"host\": \"server01\"},\"timestamp\": \"$now\",\"fields\": {\"value\": 100}}]}" -H "Content-Type: application/json" -ipv4 http://localhost:8086/write

echo "inserting data"
curl -d "{\"database\" : \"foo\", \"retentionPolicy\" : \"bar\", \"points\": [{\"name\": \"cpu\", \"tags\": {\"region\":\"uswest\",\"host\": \"server01\"},\"timestamp\": \"$tomorrow\",\"fields\": {\"value\": 200}}]}" -H "Content-Type: application/json" -ipv4 http://localhost:8086/write

sleep 1

echo "querying data"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "db=foo" --data-urlencode "q=SELECT count(value) FROM \"foo\".\"bar\".cpu"

echo "querying data"
curl -G -ipv4 http://localhost:8086/query --data-urlencode "db=foo" --data-urlencode "q=SELECT count(value) FROM \"foo\".\"bar\".cpu where time > now() - 2d and time < now() + 10d"
```

I was able to reproduce with @neonstalwart 's script...
```
[http] 2015/04/13 15:44:26 127.0.0.1 - - [13/Apr/2015:15:44:26 -0400] GET /query?db=foo&q=SELECT%20count%28value%29%20FROM%20%22foo%22.%22bar%22.cpu HTTP/1.1 200 110 - curl/7.35.0 7dfaa3c5-e215-11e4-8024-000000000000 32.623578ms
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x4dabea]

goroutine 284398 [running]:
github.com/influxdb/influxdb.(*Measurement).idsForExpr(0xc20860d560, 0xc208c9af00, 0x0, 0x0, 0x0, 0xc208ec1300, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/database.go:365 +0x91a
github.com/influxdb/influxdb.(*Measurement).walkWhereForSeriesIds(0xc20860d560, 0x7ff708325340, 0xc208c9af00, 0xc208c9b140, 0x0, 0x0, 0x0, 0xc208ec1300, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/database.go:498 +0xcdb
github.com/influxdb/influxdb.(*Measurement).walkWhereForSeriesIds(0xc20860d560, 0x7ff708325340, 0xc208c9b020, 0xc208c9b140, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/database.go:436 +0x1f6
github.com/influxdb/influxdb.(*Measurement).filters(0xc20860d560, 0xc209db60c0, 0xc208ec1540)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/database.go:292 +0xac
github.com/influxdb/influxdb.(*Measurement).tagSets(0xc20860d560, 0xc209db60c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/database.go:313 +0x67
github.com/influxdb/influxdb.(*tx).CreateMapReduceJobs(0xc20c203c80, 0xc209db60c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/tx.go:125 +0x1009
github.com/influxdb/influxdb/influxql.(*Planner).Plan(0xc209de37a0, 0xc209db60c0, 0x2710, 0x1, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/influxql/engine.go:710 +0x319
github.com/influxdb/influxdb.(*Server).planSelectStatement(0xc208062200, 0xc209db60c0, 0x2710, 0x0, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/server.go:2418 +0x14a
github.com/influxdb/influxdb.(*Server).executeSelectStatement(0xc208062200, 0x0, 0xc209db60c0, 0xc20a0cea5e, 0x3, 0x0, 0xc2086dd7a0, 0x2710, 0x0, 0x0)
	/home/dgnorton/go/src/github.com/influxdb/influxdb/server.go:2245 +0xba
github.com/influxdb/influxdb.func·039()
	/home/dgnorton/go/src/github.com/influxdb/influxdb/server.go:2150 +0xab4
created by github.com/influxdb/influxdb.(*Server).ExecuteQuery
	/home/dgnorton/go/src/github.com/influxdb/influxdb/server.go:2231 +0x314
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,7 @@
 - [#2282](https://github.com/influxdb/influxdb/pull/2282): Use "value" as field name for OpenTSDB input.
 - [#2283](https://github.com/influxdb/influxdb/pull/2283): Fix bug when restarting an entire existing cluster.
 - [#2287](https://github.com/influxdb/influxdb/pull/2287): Fix data race during SHOW RETENTION POLICIES.
+- [#2288](https://github.com/influxdb/influxdb/pull/2288): Fix expression parsing bug.
 
 ## Features
 - [#2276](https://github.com/influxdb/influxdb/pull/2276): Broker topic truncation.

--- a/influxql/parser.go
+++ b/influxql/parser.go
@@ -1737,15 +1737,44 @@ func (p *Parser) ParseExpr() (Expr, error) {
 			}
 		}
 
-		// Assign the new root based on the precendence of the LHS and RHS operators.
-		if lhs, ok := expr.(*BinaryExpr); ok && lhs.Op.Precedence() < op.Precedence() {
-			expr = &BinaryExpr{
-				LHS: lhs.LHS,
-				RHS: &BinaryExpr{LHS: lhs.RHS, RHS: rhs, Op: op},
-				Op:  lhs.Op,
+		// Find the right spot in the tree to add the new expression:
+		//   Descent the RHS of the expression tree until we reach the last
+		//   BinaryExpr OR a BinaryExpr who's RHS has an operator with
+		//   precedence >= the operator being added.
+		lhs := expr
+		isroot := true
+		for {
+			e, ok := lhs.(*BinaryExpr)
+			if !ok {
+				break
+			}
+
+			r, ok := e.RHS.(*BinaryExpr)
+			if !ok || r.Op.Precedence() >= op.Precedence() {
+				break
+			}
+
+			lhs = e.RHS
+			isroot = false
+		}
+
+		// Add the new expression to the tree.
+		if lhs2, ok := lhs.(*BinaryExpr); ok && lhs2.Op.Precedence() < op.Precedence() {
+			if isroot {
+				expr = &BinaryExpr{
+					LHS: lhs2.LHS,
+					RHS: &BinaryExpr{LHS: lhs2.RHS, RHS: rhs, Op: op},
+					Op:  lhs2.Op,
+				}
+			} else {
+				lhs2.RHS = &BinaryExpr{LHS: lhs2.RHS, RHS: rhs, Op: op}
 			}
 		} else {
-			expr = &BinaryExpr{LHS: expr, RHS: rhs, Op: op}
+			if isroot {
+				expr = &BinaryExpr{LHS: lhs, RHS: rhs, Op: op}
+			} else {
+				lhs = &BinaryExpr{LHS: lhs, RHS: rhs, Op: op}
+			}
 		}
 	}
 }

--- a/influxql/parser.go
+++ b/influxql/parser.go
@@ -1739,17 +1739,17 @@ func (p *Parser) ParseExpr() (Expr, error) {
 			}
 		}
 
-		// Find the right spot in the tree to add the new expression:
-		//   Descent the RHS of the expression tree until we reach the last
-		//   BinaryExpr OR a BinaryExpr who's RHS has an operator with
-		//   precedence >= the operator being added.
+		// Find the right spot in the tree to add the new expression by
+		// descending the RHS of the expression tree until we reach the last
+		// BinaryExpr or a BinaryExpr whose RHS has an operator with
+		// precedence >= the operator being added.
 		for node := root; ; {
 			r, ok := node.RHS.(*BinaryExpr)
 			if !ok || r.Op.Precedence() >= op.Precedence() {
+				// Add the new expression here and break.
 				node.RHS = &BinaryExpr{LHS: node.RHS, RHS: rhs, Op: op}
 				break
 			}
-
 			node = r
 		}
 	}

--- a/influxql/parser.go
+++ b/influxql/parser.go
@@ -1706,10 +1706,12 @@ func (p *Parser) parseVarRef() (*VarRef, error) {
 // ParseExpr parses an expression.
 func (p *Parser) ParseExpr() (Expr, error) {
 	var err error
+	// Dummy root node.
+	root := &BinaryExpr{}
 
 	// Parse a non-binary expression type to start.
 	// This variable will always be the root of the expression tree.
-	expr, err := p.parseUnaryExpr()
+	root.RHS, err = p.parseUnaryExpr()
 	if err != nil {
 		return nil, err
 	}
@@ -1720,7 +1722,7 @@ func (p *Parser) ParseExpr() (Expr, error) {
 		op, _, _ := p.scanIgnoreWhitespace()
 		if !op.isOperator() {
 			p.unscan()
-			return expr, nil
+			return root.RHS, nil
 		}
 
 		// Otherwise parse the next expression.
@@ -1741,40 +1743,14 @@ func (p *Parser) ParseExpr() (Expr, error) {
 		//   Descent the RHS of the expression tree until we reach the last
 		//   BinaryExpr OR a BinaryExpr who's RHS has an operator with
 		//   precedence >= the operator being added.
-		lhs := expr
-		isroot := true
-		for {
-			e, ok := lhs.(*BinaryExpr)
-			if !ok {
-				break
-			}
-
-			r, ok := e.RHS.(*BinaryExpr)
+		for node := root; ; {
+			r, ok := node.RHS.(*BinaryExpr)
 			if !ok || r.Op.Precedence() >= op.Precedence() {
+				node.RHS = &BinaryExpr{LHS: node.RHS, RHS: rhs, Op: op}
 				break
 			}
 
-			lhs = e.RHS
-			isroot = false
-		}
-
-		// Add the new expression to the tree.
-		if lhs2, ok := lhs.(*BinaryExpr); ok && lhs2.Op.Precedence() < op.Precedence() {
-			if isroot {
-				expr = &BinaryExpr{
-					LHS: lhs2.LHS,
-					RHS: &BinaryExpr{LHS: lhs2.RHS, RHS: rhs, Op: op},
-					Op:  lhs2.Op,
-				}
-			} else {
-				lhs2.RHS = &BinaryExpr{LHS: lhs2.RHS, RHS: rhs, Op: op}
-			}
-		} else {
-			if isroot {
-				expr = &BinaryExpr{LHS: lhs, RHS: rhs, Op: op}
-			} else {
-				lhs = &BinaryExpr{LHS: lhs, RHS: rhs, Op: op}
-			}
+			node = r
 		}
 	}
 }

--- a/influxql/parser_test.go
+++ b/influxql/parser_test.go
@@ -1172,6 +1172,32 @@ func TestParser_ParseExpr(t *testing.T) {
 			},
 		},
 
+		// Complex binary expression.
+		{
+			s: `time > now() - 1d AND time < now() + 1d`,
+			expr: &influxql.BinaryExpr{
+				Op: influxql.AND,
+				LHS: &influxql.BinaryExpr{
+					Op:  influxql.GT,
+					LHS: &influxql.VarRef{Val: "time"},
+					RHS: &influxql.BinaryExpr{
+						Op:  influxql.SUB,
+						LHS: &influxql.Call{Name: "now"},
+						RHS: &influxql.DurationLiteral{Val: mustParseDuration("1d")},
+					},
+				},
+				RHS: &influxql.BinaryExpr{
+					Op:  influxql.LT,
+					LHS: &influxql.VarRef{Val: "time"},
+					RHS: &influxql.BinaryExpr{
+						Op:  influxql.ADD,
+						LHS: &influxql.Call{Name: "now"},
+						RHS: &influxql.DurationLiteral{Val: mustParseDuration("1d")},
+					},
+				},
+			},
+		},
+
 		// Function call (empty)
 		{
 			s: `my_func()`,
@@ -1375,18 +1401,14 @@ func BenchmarkParserParseStatement(b *testing.B) {
 // MustParseSelectStatement parses a select statement. Panic on error.
 func MustParseSelectStatement(s string) *influxql.SelectStatement {
 	stmt, err := influxql.NewParser(strings.NewReader(s)).ParseStatement()
-	if err != nil {
-		panic(err.Error())
-	}
+	panicIfErr(err)
 	return stmt.(*influxql.SelectStatement)
 }
 
 // MustParseExpr parses an expression. Panic on error.
 func MustParseExpr(s string) influxql.Expr {
 	expr, err := influxql.NewParser(strings.NewReader(s)).ParseExpr()
-	if err != nil {
-		panic(err.Error())
-	}
+	panicIfErr(err)
 	return expr
 }
 
@@ -1420,8 +1442,18 @@ func newAlterRetentionPolicyStatement(name string, DB string, d time.Duration, r
 // mustMarshalJSON encodes a value to JSON.
 func mustMarshalJSON(v interface{}) []byte {
 	b, err := json.Marshal(v)
-	if err != nil {
-		panic("marshal json: " + err.Error())
-	}
+	panicIfErr(err)
 	return b
+}
+
+func mustParseDuration(s string) time.Duration {
+	d, err := influxql.ParseDuration(s)
+	panicIfErr(err)
+	return d
+}
+
+func panicIfErr(err error) {
+	if err != nil {
+		panic(err)
+	}
 }

